multi: add spend pruner.

dnldd opened the pull request:
This adds a spend journal pruner to manage the removal of spend journal entries taking into consideration the consumers of such entries. Associated tests have been added.

dnldd commented:
@davecgh I'm unsure of where the spend pruner should be, can it be under internal or do you have a better location for it.

This is currently missing persisting dependency entries, I'm looking for feedback however so please do take a look as well @rstaudt2.

davecgh commented:
I think it's fine under internal.  Nothing external to dcrd should need to access it.

matheusd commented on blockchain/internal/spendpruner/errors.go:
```diff
@@ -0,0 +1,50 @@
+// Copyright (c) 2019-2020 The Decred developers
```
Should be 2019-2021 as well

matheusd reviewed (commented):
Can you give a brief high level overview on what exactly this aims to achieve in the medium/long term and on the design for this pruner subsystem?

I _think_ I see where this is going, but wanted to make sure I understand the overall picture

rstaudt2 reviewed (commented):
Agreed with @matheusd that an overview would be helpful.  I'll do my best to summarize in the meantime based on prior conversations.

This need originates from the PR the introduces asynchronous indexing https://github.com/decred/dcrd/pull/2219#discussion_r604539863:

> My understanding of the problem that is trying to be solved here is:
> 
>  - We are adding an asynchronous consumer of block connect/disconnect notifications (possibly multiple consumers in the future)
>  - The consumer needs the spend journal entry for disconnected blocks, but currently, that is removed immediately when disconnecting a block and is therefore not available
> 
> Given that, I think the goal of an ideal solution for this would be:
> 
>  - The spend journal is retained only while there are still consumers that need it

This my high-level understanding of the `SpendJournalPruner` design introduced in this PR, which mostly makes sense to me, with a few open questions on how certain scenarios will be handled:

#### On `dcrd` startup:
- A single instance of `SpendJournalPruner` is created at startup
  - Each `SpendConsumer` will register with the `SpendJournalPruner` using the `AddConsumer(consumer SpendConsumer)` function at startup
  - `SpendConsumer` represents an entity that doesn't want spend journal entries removed before they confirm that they no longer need it (the initial use case being asynchronous indexers)

#### On block disconnect:
- `BlockChain` no longer removes the spend journal entries itself and instead
  calls the `SpendJournalPruner` function `AddDependents(hash *chainhash.Hash)`
  after disconnecting a block
  - `AddDependents(hash *chainhash.Hash)` iterates through each registered `SpendConsumer`, checking if it needs the spend journal entry associated with the block hash.  `SpendJournalPruner` tracks the consumers that depend on the spend journal entry in its internal `dependents` map

#### When a consumer is done with a spend journal entry:
- When a consumer is done with a spend journal entry, it will call the `SpendJournalPruner` function `RemoveDependency(hash *chainhash.Hash, consumerID string)`
- `RemoveDependency(hash *chainhash.Hash, consumerID string)` will remove the dependency on the spend journal entry for the provider consumer id.  If no more dependencies exist, then the spend journal entry will be removed.

Design questions:
- If there are no consumers (i.e., indexes disabled), then when would the spend journal entry get removed? (`RemoveDependency` will never be called).
- When restarting `dcrd` with a non-empty `dependents` map, how does it get restarted on startup?  Is it expected that each consumer track this and call `AddDependents` on startup?  If that is the case, won't this be inefficient since each consumer is calling `AddDependents` which in turn checks if each registered consumer needs the spend journal entry?

rstaudt2 commented on blockchain/internal/spendpruner/pruner.go:
```diff
@@ -0,0 +1,160 @@
+
+// UpdateConsumerTip updates the tip hash of the associated spend journal
+// consumer with the provided id.
+func (s *SpendJournalPruner) UpdateConsumerTip(id string, tipHash *chainhash.Hash) error {
```
Where/when would `UpdateConsumerTip` be called, and how does this fit into the overall picture?  It seems like it would be up to the consumers to manage this and not the `SpendJournalPruner` since the `SpendJournalPruner` is responsible for removing spend journal entries but not keeping consumers in sync.

dnldd replied:
`UpdateConsumerTip` will be called when the represented consumer's tip hash is updated. Considering the pruner keeps a list of consumers I think it's best it provides an update function for the represented consumers to sync their updated state.

rstaudt2 replied:
I might be missing something else, but I still am not following why the `SpendJournalPruner` would be updating the tip on consumers.  The `SpendJournalPruner` is just managing the spend journal removal, not keeping the consumers in sync.  Further, it is not even tracking the tip or the chain itself, it is just tracking spend journal entries that are pending removal.

dnldd replied:
The `SpendJournalPruner` will not b updating the tip of consumers. It holds references of those consumers and `UpdateConsumerTip` provides access to updating those references. `UpdateConsumerTip` will be called by the consumers, for example for the indexers, when a new block is connected `UpdateConsumerTip` will be called to sync the state of the consumer with the reference the pruner has.

rstaudt2 replied:
> UpdateConsumerTip will be called by the consumers, for example for the indexers, when a new block is connected UpdateConsumerTip will be called to sync the state of the consumer with the reference the pruner has.

I'm not clear on why this would be the responsibility of the `SpendJournalPruner`.  Is the `SpendJournalPruner` only intended to prune spend journal entries, or is it also syncing the state of consumers?  These seem like two separate things, and it's hard to tell how this function fits in since it's only used in the test.

dnldd replied:
Will start integrating this into the async indexers work, think that'd best show my intentions for `UpdateConsumerTip` or further refine the `SpendConsumer` interface.

dnldd commented on blockchain/internal/spendpruner/pruner.go:
```diff
@@ -0,0 +1,160 @@
+		pruneCh:    make(chan *chainhash.Hash),
+	}
+
+	go prunefilter.handleSpendPrunes(ctx)
```
Noted, updating.

dnldd commented:
Thanks for the design outline @rstaudt2, I think you captured all the details. 

> When restarting dcrd with a non-empty dependents map, how does it get restarted on startup? Is it expected that each consumer track this and call AddDependents on startup? If that is the case, won't this be inefficient since each consumer is calling AddDependents which in turn checks if each registered consumer needs the spend journal entry?

The current state of this PR is missing the persistence aspect of the pruner. Will be working that in next. On startup the pruner will read existing dependencies from the database and persist them on shutdown.

> If there are no consumers (i.e., indexes disabled), then when would the spend journal entry get removed? (RemoveDependency will never be called).

Good catch, I plan on updating `AddDependents` to signal the prune process with the provided block hash if there are no dependencies generated for it. Perhaps a better name for that will now be `MaybeAddDependents`.

dnldd commented on blockchain/internal/spendpruner/pruner.go:
```diff
@@ -0,0 +1,160 @@
+// SpendJournalPruner represents a spend journal data pruner that ensures
+// data needed by spend consumers are retained until no longer needed.
+type SpendJournalPruner struct {
+	queryer    indexers.ChainQueryer
```
Agreed on both suggestions. The consumer's queryer is a subset of what `indexers.ChainQueryer` requires, shouldn't be an indexer interface.

matheusd commented:
Thank you @rstaudt2 for the info. Overall, seems reasonable. So far, I only have a couple of suggestions:

> BlockChain no longer removes the spend journal entries itself and instead
calls the SpendJournalPruner function AddDependents(hash *chainhash.Hash)
after disconnecting a block 

`AddDependents` seems like a too generic name to me and made it harder to understand the purpose of code on a first read ("what's being dependent on and for what reasons" were the first things I thought upon reading this that weren't immediately clear). Some ideas to improve (though feel free to wait for other ppl to chime in to gather more feedback before implementing these):

Since this is meant to replace the very last step when disconnecting a block, I suggest treating the block as effectively disconnected so `AddDependents` woud become `DisconnectBlock` (or something to that effect), `RemoveDependency` would become `SpendDataConsumed` (to signal the consumer is done _consuming_ the spend data and the pruner can carry on) and `DependencyExists` would become `NeedsSpendData`.  

Also, various places just refer to "hash" as in:

> RemoveDependency removes the provided consumer dependency hash from the spend pruner.

Which doesn't make it immediately obvious if you're talking about a transaction, block or some other hash (if one is only looking at this function's doc without keeping in mind the larger scope). So I suggest replacing "hash" in such docs for the concrete type (e.g. "... consumer dependency block ..." or "... consumer dependency block hash"). I see you've already done that on some docs, and this can be left as an improvement for later stages in the PR.

On the design: based on @rstaudt2 info, the pruner is meant to run asynchronously, right? I guess the async bits still need to be filled in, because the pruner itself has no `Run()` or similar method (or maybe `AddDepdents` is meant to be run as a goroutine)?

If the pruner is meant to run as an async service, can this impact async indexers in case of fast reorgs? For example:

```
b1 -> b2  -> b3  -> b4
  \-> b2a -> b3a

Reception order: b1, b2, b2a, b3a, b3, b4
```

Notice `b2` is connected once, then disconnected, then reconnected. Is the pruner required to safely keep track that `b2` was reconnected if it hasn't pruned `b2` yet (because some consumer took a long time to call `RemoveDependency`)?  Mostly I'm trying to understand who has the responsibility for maintaining consistency of the spend journal (whether it's the pruner, the blockchain code or the client consumers).

dnldd commented:
Thanks for the feedback @matheusd,  will be pushing an update later today which should clear up most of the issues being discussed as well as answer any remaining questions.

dnldd commented:
> AddDependents seems like a too generic name to me and made it harder to understand the purpose of code on a first read ("what's being dependent on and for what reasons" were the first things I thought upon reading this that weren't immediately clear). Some ideas to improve (though feel free to wait for other ppl to chime in to gather more feedback before implementing these):

`AddDependents` (renamed to `AddDependencies`) only creates spend consumer dependencies for the spend data associated with the block hash passed. These dependencies serve as records for the pruner to keep track of what still needs the spend data 
and prevent its removal if its still needed.

> Since this is meant to replace the very last step when disconnecting a block, I suggest treating the block as effectively disconnected so AddDependents woud become DisconnectBlock (or something to that effect), RemoveDependency would become SpendDataConsumed (to signal the consumer is done consuming the spend data and the pruner can carry on) and DependencyExists would become NeedsSpendData.

Hmm `NeedsSpendData` is already part of the `SpendConsumer` interface and I don't think it maps well in meaning to  the current state of `DependencyExists`. `DependencyExists` primarily serves as a lookup for dependencies created for spend data associated with the passed block hash. Same for `AddDependents` and `DisconnectBlock`, their meanings do not map well. Do take another look with the updates made.

> Notice b2 is connected once, then disconnected, then reconnected. Is the pruner required to safely keep track that b2 was reconnected if it hasn't pruned b2 yet (because some consumer took a long time to call RemoveDependency)?  Mostly I'm trying to understand who has the responsibility for maintaining consistency of the spend journal (whether it's the pruner, the blockchain code or the client consumers).

The pruner will not remove a spend entry if the entry still has dependencies referencing it. `b2` entry in the spend pruner will remain on the first addition if one of the consumers take to long to call `RemoveDependency`, if the `b2` is reconnected by then and the more consumers need it then their dependency entries will be added. Ultimately it's the consumer that determines how quickly the spend journal updates through their `RemoveDependency` updates.

rstaudt2 commented:
> Since this is meant to replace the very last step when disconnecting a block, I suggest treating the block as effectively disconnected so AddDependents would become DisconnectBlock (or something to that effect), RemoveDependency would become SpendDataConsumed (to signal the consumer is done consuming the spend data and the pruner can carry on) and DependencyExists would become NeedsSpendData.

I like this suggestion and do think it would be easier to follow. 

> Notice b2 is connected once, then disconnected, then reconnected. Is the pruner required to safely keep track that b2 was reconnected if it hasn't pruned b2 yet (because some consumer took a long time to call RemoveDependency)? Mostly I'm trying to understand who has the responsibility for maintaining consistency of the spend journal (whether it's the pruner, the blockchain code or the client consumers).

This is a great point too, and I'd add a test to ensure it behaves as suspected.  If a block gets reconnected, `BlockChain` will probably need to tell `SpendJournalPruner` that it was reconnected so that if that block's spend journal entry was currently pending removal, it knows that it should no longer remove it.

rstaudt2 commented on blockchain/chain.go:
```diff
@@ -1683,6 +1690,26 @@ func (b *BlockChain) MainChainHasBlock(hash *chainhash.Hash) bool {
+// provided hash.
+//
+// This function is safe for concurrent access.
+func (b *BlockChain) RemoveSpendEntry(hash *chainhash.Hash) error {
```
I agree that it's misleading and potentially dangerous to have this on the public `BlockChain` API since it shouldn't just be used by any caller.  My original suggestion was to have `SpendJournalEntry` accept this as a function type on creation, in which case it would not need to be exposed like this.

dnldd replied:
I think the idea of having `internalChainAdapter` is better overall for this because the pruner will still need access to other blockchain functions. Would be better to do `NewSpendJournalPruner(chainAdapter)` than `NewSpendJournalPruner(chainAdapter, removeSpendEntry)` considering `removeSpendEntry` could be part of the adapter.

dnldd commented on blockchain/chain.go:
```diff
@@ -1683,6 +1690,26 @@ func (b *BlockChain) MainChainHasBlock(hash *chainhash.Hash) bool {
+// block at the provided height.
+//
+// This function is safe for concurrent access.
+func (b *BlockChain) Ancestor(block *chainhash.Hash, height int64, hash *chainhash.Hash) bool {
```
Maintaining the name since `Is..` would indicate the returns a bool and returning the ancestor block hash instead.

rstaudt2 commented on blockchain/internal/spendpruner/pruner.go:
```diff
@@ -0,0 +1,292 @@
+
+			// Remove the prune entry of the provided connected block hash
+			// and its consumer dependencies.
+			s.removeSpendEntry(blockHash)
```
If I'm reading this correctly, this is removing a spend journal entry when a block is connected?  The spend journal entry should only be removed after `BlockChain` _disconnects_ a block (and consumers no longer need it).

dnldd replied:
Yeah that was a bad function name to use there.  It's actually only removing the key/value pair of the spend consumer dependencies and the associated connected block from the prune set since block connection invalidates the need to prune the spend journal entry of the block. I've renamed the function to `removeSpendConsumerDeps`, good catch.

rstaudt2 replied:
Got it, thanks for clarifying.  Renaming it definitely helps.  Planning on reviewing the latest changes more in depth soon yet just noticed this when taking a quick look.

JoeGruffins reviewed (commented):
Some inline comments.

dnldd commented on blockchain/internal/spendpruner/db.go:
```diff
@@ -0,0 +1,85 @@
+func serializeSpendConsumerDeps(deps []string) []byte {
+	buf := new(bytes.Buffer)
+	for idx := 0; idx < len(deps); idx++ {
+		buf.WriteString(deps[0])
```
good catch.

rstaudt2 commented on blockchain/internal/spendpruner/pruner.go:
```diff
@@ -0,0 +1,263 @@
+		dependents = append(dependents, consumer.ID())
+		s.dependents[*blockHash] = dependents
+		s.dependentsMtx.Unlock()
+
```
Extra newline here.

rstaudt2 commented on blockchain/internal/spendpruner/pruner.go:
```diff
@@ -0,0 +1,263 @@
+	// block hash.
+	err := s.db.Update(func(tx database.Tx) error {
+		return dbUpdateSpendConsumerDeps(tx, *blockHash, nil)
+	})
```
Is it intended that in this case, the db is updated after deleting `blockHash` from `s.dependents`, but the db is not updated in `RemoveSpendConsumerDependency`?  It appears that it might be useful for `RemoveSpendConsumerDependency` to use `removeSpendConsumerDeps` so that the removal logic is in one place.

dnldd replied:
Good catch, `RemoveSpendConsumerDependency` needs to persist to the db after the removal. `removeSpendConsumerDeps` wipes all the spend consumer deps for the provided blockhash and updates the db when done while `RemoveSpendConsumerDependency` only wipes the spend consumer set for the provided blockhash on condition the last entry was just removed. For that reason I don't think calling `removeSpendConsumerDeps` from `RemoveSpendConsumerDependency` to keep removal logic in one place would be ideal since `removeSpendConsumerDeps` would have to be updated to cover the two contexts described.

rstaudt2 commented on blockchain/chain.go:
```diff
@@ -1069,6 +1072,13 @@ func (b *BlockChain) reorganizeChainInternal(target *blockNode) error {
+		err = b.spendPruner.AddSpendConsumerDeps(block.Hash())
+		if err != nil {
+			return err
+		}
```
Could this be handled internally within `MaybePruneSpendData`? (i.e., `MaybePruneSpendData` would first add any dependencies and then only remove the spend journal entry if there are no dependencies).  I think it would simplify the flow unless there is a reason why they need to be separate.

dnldd replied:
It could but it'd be less readable, I think it's beneficial to distinguish between the two processes, don't think there is much gained combining them under `MaybePruneSpendData`.

rstaudt2 replied:
You could still have a separate function for adding dependencies within the `SpendJournalPruner`, but I don't see why `BlockChain` needs to know about that.  Generally, fewer entry points between components leads to better encapsulation and more readable code.  Also, if you do decide to keep them separate, it seems that `AddSpendConsumerDeps` can be called immediately before `MaybePruneSpendData` rather than having them in two separate places, correct?

dnldd replied:
Yes you're correct, updating.

dnldd commented on blockchain/internal/spendpruner/interface.go:
```diff
@@ -0,0 +1,30 @@
+	ID() string
+
+	// UpdateTip sets the tip of the consumer to the provided block hash.
+	UpdateTip(hash *chainhash.Hash)
```
I think that should work, will make those changes and provide feedback.

dnldd commented on blockchain/upgrade.go:
```diff
@@ -3183,6 +3183,49 @@ func upgradeToVersion10(ctx context.Context, db database.DB, dbInfo *databaseInf
+	return db.Update(func(dbTx database.Tx) error {
+		meta := dbTx.Metadata()
+		_, err := meta.CreateBucketIfNotExists(spendConsumerDepsBucketName)
+		return err
```
Unsure about this, will wait for more feedback before making any changes.

davecgh replied:
Agreed.  Typically if it's just a new bucket, I tend to do the following:

1) Create the bucket in the startup path if it doesn't already exist
2) At some later point, when there is a database version migration needed for something due to actual changed records:
a) Create the bucket when initializing a new database
b) Roll the conditional bucket creation into that upgrade and remove it from the startup path

rstaudt2 reviewed (changes requested):
Overall this is looking pretty good.  I have called out a few additional things inline.  I've also been testing this, and it seems to be working properly for the most part, but I see a panic in one scenario:

- Sync the chain with indexes enabled (I was using `--txindex` and `--addrindex`)
- Restart the chain with indexes disabled
- Roll a few blocks back using `dcrctl invalidateblock`
- Restart the chain with indexes enabled
- The indexing code will panic trying to recover from spend journal entries that no longer exist

dnldd commented:
Resolved the panic, will sort out the rest of the review issues and update the PRs, good catch 👍 

```sh
2021-07-15 00:01:36.411 [DBG] CHAN: Difficulty retarget at block height 24                                                                                                                              
2021-07-15 00:01:36.411 [DBG] CHAN: Old target 206eeeee (6eeeee0000000000000000000000000000000000000000000000000000000000)                                                                              
2021-07-15 00:01:36.411 [DBG] CHAN: New target 207fffff (7fffff0000000000000000000000000000000000000000000000000000000000)                                                                              
2021-07-15 00:01:36.411 [DBG] MINR: Created new block template (1 transactions, 1 stake transactions, 0 treasury transactions, 0 in fees, 1 signature operations, 323 bytes, target difficulty 7fffff000
0000000000000000000000000000000000000000000000000000000, stake difficulty 0.0002)                                                                                                                       
2021-07-15 00:01:37.036 [DBG] CHAN: Difficulty retarget at block height 24                                                                                                                              
2021-07-15 00:01:37.036 [DBG] CHAN: Old target 206eeeee (6eeeee0000000000000000000000000000000000000000000000000000000000)                                                                              
2021-07-15 00:01:37.036 [DBG] CHAN: New target 207fffff (7fffff0000000000000000000000000000000000000000000000000000000000)                                                                              
2021-07-15 00:01:37.180 [DBG] FEES: Updated moving averages into block 24
2021-07-15 00:01:37.181 [INF] MINR: Block submitted via CPU miner accepted (hash 371ad9189e7f8b8bc780a341c649e78d2093bd088c1c3993196d5163b65ccd02, height 24, amount 300 DCR)
2021-07-15 00:01:37.181 [DBG] MINR: Considering 0 transactions for inclusion to new block
2021-07-15 00:01:37.181 [DBG] MINR: Created new block template (1 transactions, 1 stake transactions, 0 treasury transactions, 0 in fees, 1 signature operations, 323 bytes, target difficulty 7fffff0000000000000000000000000000000000000000000000000000000000, stake difficulty 0.0002)
2021-07-15 00:01:37.932 [DBG] FEES: Updated moving averages into block 25
2021-07-15 00:01:37.933 [INF] MINR: Block submitted via CPU miner accepted (hash 4f89ae556a2d637fe375d7ecc18831378a1f7c467076b123caf40dabd8182d7f, height 25, amount 300 DCR)
2021-07-15 00:01:37.933 [DBG] MINR: Considering 0 transactions for inclusion to new block
2021-07-15 00:01:37.933 [DBG] MINR: Created new block template (1 transactions, 1 stake transactions, 0 treasury transactions, 0 in fees, 1 signature operations, 323 bytes, target difficulty 7fffff0000000000000000000000000000000000000000000000000000000000, stake difficulty 0.0002)
2021-07-15 00:02:01.164 [INF] CHAN: Rolling the chain back to block 708b671e6a76032c208485fb9a81d7c0b9a516d8ac9cf3e769b223d813e88f93 (height 13) due to manual invalidation
2021-07-15 00:02:02.693 [INF] CHAN: REORGANIZE: Chain forks at 708b671e6a76032c208485fb9a81d7c0b9a516d8ac9cf3e769b223d813e88f93 (height 13)
2021-07-15 00:02:02.693 [INF] CHAN: REORGANIZE: Old best chain tip was 4f89ae556a2d637fe375d7ecc18831378a1f7c467076b123caf40dabd8182d7f (height 25)
2021-07-15 00:02:02.693 [INF] CHAN: REORGANIZE: New best chain tip is 708b671e6a76032c208485fb9a81d7c0b9a516d8ac9cf3e769b223d813e88f93 (height 13)
2021-07-15 00:02:02.693 [DBG] CHAN: Chain latched to current at block 708b671e6a76032c208485fb9a81d7c0b9a516d8ac9cf3e769b223d813e88f93 (height 13)
2021-07-15 00:02:02.693 [DBG] MINR: Considering 0 transactions for inclusion to new block
2021-07-15 00:02:02.694 [DBG] MINR: Created new block template (1 transactions, 1 stake transactions, 0 treasury transactions, 0 in fees, 1 signature operations, 323 bytes, target difficulty 7fffff0000000000000000000000000000000000000000000000000000000000, stake difficulty 0.0002)
^C2021-07-15 00:02:34.017 [INF] DCRD: Received signal (interrupt).  Shutting down...
2021-07-15 00:02:34.018 [WRN] SRVR: Server shutting down
2021-07-15 00:02:34.018 [INF] INDX: Index subscriber shutting down
2021-07-15 00:02:34.018 [INF] AMGR: Address manager shutting down
2021-07-15 00:02:34.018 [WRN] RPCS: RPC server shutting down
2021-07-15 00:02:34.018 [INF] RPCS: RPC server shutdown complete
2021-07-15 00:02:34.019 [DBG] CHAN: UTXO cache flush starting (3 entries, 0.00 MiB (0.00%), 18.42% hit ratio, height: 13)
2021-07-15 00:02:34.127 [DBG] CHAN: UTXO cache flush completed (3 entries flushed, 0 entries remaining, 0.00 MiB (0.00%))
2021-07-15 00:02:34.127 [INF] SRVR: Server shutdown complete
2021-07-15 00:02:34.127 [INF] DCRD: Gracefully shutting down the UTXO database...
2021-07-15 00:02:34.127 [INF] DCRD: Gracefully shutting down the block database...
2021-07-15 00:02:34.146 [INF] DCRD: Shutdown complete

~/sh/node 1m 7s
❯ dcrd --rpcuser=user --rpcpass=pass --miningaddr=SsWKp7wtdTZYabYFYSc9cnxhwFEjA5g4pFc --profile=127.0.0.1:2555 --appdata=/Users/dnldd/sh/node --debuglevel=debug --txindex --addrindex --simnet
2021-07-15 00:02:40.114 [WRN] DCRD: open /Users/dnldd/sh/node/dcrd.conf: no such file or directory
2021-07-15 00:02:40.114 [INF] DCRD: Version 1.7.0-pre (Go version go1.16.5 darwin/arm64)
2021-07-15 00:02:40.114 [INF] DCRD: Home dir: /Users/dnldd/sh/node
2021-07-15 00:02:40.114 [INF] DCRD: Loading block database from '/Users/dnldd/sh/node/data/simnet/blocks_ffldb'
2021-07-15 00:02:40.114 [INF] DCRD: Creating profiling server listening on 127.0.0.1:2555
2021-07-15 00:02:40.173 [INF] DCRD: Block database loaded
2021-07-15 00:02:40.173 [INF] CHAN: Loading UTXO database from '/Users/dnldd/sh/node/data/simnet/utxodb'
2021-07-15 00:02:40.231 [INF] CHAN: UTXO database loaded
2021-07-15 00:02:40.311 [DBG] FEES: Loaded fee estimator database
2021-07-15 00:02:40.315 [INF] CHAN: Loading block index...
2021-07-15 00:02:40.316 [DBG] CHAN: Block index loaded in 521.292µs
2021-07-15 00:02:40.316 [INF] STKE: Stake database version 1 loaded
2021-07-15 00:02:40.316 [INF] CHAN: UTXO cache initializing (max size: 150 MiB)...
2021-07-15 00:02:40.316 [INF] CHAN: UTXO cache initialization completed
2021-07-15 00:02:40.316 [INF] CHAN: Blockchain database version info: chain: 10, compression: 1, block index: 3, spend journal: 3
2021-07-15 00:02:40.316 [INF] CHAN: UTXO database version info: version: 1, compression: 1, utxo set: 3
2021-07-15 00:02:40.316 [INF] CHAN: Best known header: height 13, hash 708b671e6a76032c208485fb9a81d7c0b9a516d8ac9cf3e769b223d813e88f93
2021-07-15 00:02:40.316 [INF] CHAN: Chain state: height 13, hash 708b671e6a76032c208485fb9a81d7c0b9a516d8ac9cf3e769b223d813e88f93, total transactions 26, work 28, progress 100.00%
2021-07-15 00:02:40.316 [INF] INDX: Transaction index is enabled
2021-07-15 00:02:40.316 [INF] INDX: transaction index: recovering from tip 20 (44404ebf9ca24492e7941b9435b88baba2d4383aadf912e426a83519075deab2)
2021-07-15 00:02:40.317 [INF] INDX: transaction index: index recovered to tip 13 (708b671e6a76032c208485fb9a81d7c0b9a516d8ac9cf3e769b223d813e88f93)
2021-07-15 00:02:40.317 [DBG] INDX: Current internal block ID: 13
2021-07-15 00:02:40.317 [INF] INDX: Address index is enabled
2021-07-15 00:02:40.318 [INF] INDX: address index: recovering from tip 20 (44404ebf9ca24492e7941b9435b88baba2d4383aadf912e426a83519075deab2)
2021-07-15 00:02:40.318 [WRN] INDX: Missing input 0000000000000000000000000000000000000000000000000000000000000000:4294967295:0 for tx 648797536a2b8d60a32847a31058cb3b8b138afdd1baef0c7107629e19be10e0 
while indexing block 44404ebf9ca24492e7941b9435b88baba2d4383aadf912e426a83519075deab2 (height 20)

2021-07-15 00:02:40.318 [WRN] INDX: Missing input 0000000000000000000000000000000000000000000000000000000000000000:4294967295:0 for tx 4e76eb340363020f736c5e6c583d0cb52f5292fac68ac3d10edeae3efe3eda84 
while indexing block 2f16af0696ce4759bba761b10e85b06c2ae507ba2bdbebdc198465836b1a073d (height 19)

2021-07-15 00:02:40.318 [WRN] INDX: Missing input 0000000000000000000000000000000000000000000000000000000000000000:4294967295:0 for tx 85a1582075298b115db28508e7e2a8659bc824fa381c45fac8cf237555837c0a 
while indexing block 1178498d87f345fcb82610de8112452dcda0ebfe6839b58232a8c6199b435cbb (height 18)

2021-07-15 00:02:40.318 [WRN] INDX: Missing input 0000000000000000000000000000000000000000000000000000000000000000:4294967295:0 for tx 868333ae0323fa9ab1975647d8a27b01362918bf8f271783007c4c237b59ca91 
while indexing block 0179d605483ee99d1b1c28073451025c5ef03ef20992e65a6869c19eaf26e18c (height 17)

2021-07-15 00:02:40.318 [WRN] INDX: Missing input 0000000000000000000000000000000000000000000000000000000000000000:4294967295:0 for tx 4639f463baff92446aef7b7508c470be4ed662fba5cd79b86096e78b19b88b92 
while indexing block 1d9dff301b6cc899153eb61eb8a775564827eca2f30646ded42a94233f82f342 (height 16)

2021-07-15 00:02:40.318 [WRN] INDX: Missing input 0000000000000000000000000000000000000000000000000000000000000000:4294967295:0 for tx 2cd34001aa0818afbc67a71812596a0dbff1d245830b22153f42942870d799d7 
while indexing block 2ec84f44104bfb49574e7a290ed862aec8380c1f96497123bc4e0c9ae6b4c4b2 (height 15)

2021-07-15 00:02:40.318 [WRN] INDX: Missing input 0000000000000000000000000000000000000000000000000000000000000000:4294967295:0 for tx 8cb487318372570b1d0dee375e59246062e74aded1d80d3b0ecc5d59a2983135 
while indexing block 782185629c71a1a706ec5e378d6d00a2af96ce7c702fc8fbbd47ba0511fb1471 (height 14)

2021-07-15 00:02:40.318 [INF] INDX: address index: index recovered to tip 13 (708b671e6a76032c208485fb9a81d7c0b9a516d8ac9cf3e769b223d813e88f93)
2021-07-15 00:02:40.318 [INF] INDX: Exists address index is enabled
```

dnldd commented:
Will squash the resolve commit once all issues are resolved.

rstaudt2 commented:
@dnldd Did you commit the fix that you had for the panic?  I still see it with the same reproduction steps and didn't see anything in the diff that looks like it would address it, so maybe you forget to include it or fixed a different issue?

dnldd commented:
Yeah I did, the fix is in #2219. Do test with that.

rstaudt2 commented:
> Yeah I did, the fix is in #2219. Do test with that.

Are you able to apply the fix here as well?  Otherwise, we can't really review and merge this PR independently (which I thought was the plan and why I was reviewing this PR before #2219).

dnldd commented:
Let me take another look and update you in a couple of hours. I initially thought since this PR wasn't updating indexes the panic wouldn't be originating from here. Will confirm and update ASAP.

dnldd commented:
I'm getting an error, not a panic following the steps provided. If you can provide the output you're getting that would be great. The address index expects to to be disconnecting blocks from it's index tip. It is an optional index however. Looking into it. 

```sh
dcrd --rpcuser=user --rpcpass=pass --miningaddr=SsWKp7wtdTZYabYFYSc9cnxhwFEjA5g4pFc --profile=127.0.0.1:2555 --appdata=/Users/dnldd/sh/node --debuglevel=debug --txindex --addrindex --simnet
2021-07-18 10:59:33.276 [WRN] DCRD: open /Users/dnldd/sh/node/dcrd.conf: no such file or directory
2021-07-18 10:59:33.276 [INF] DCRD: Version 1.7.0-pre (Go version go1.16.5 darwin/arm64)
2021-07-18 10:59:33.276 [INF] DCRD: Home dir: /Users/dnldd/sh/node
2021-07-18 10:59:33.276 [INF] DCRD: Creating profiling server listening on 127.0.0.1:2555
2021-07-18 10:59:33.276 [INF] DCRD: Loading block database from '/Users/dnldd/sh/node/data/simnet/blocks_ffldb'
2021-07-18 10:59:33.338 [INF] DCRD: Block database loaded
2021-07-18 10:59:33.338 [INF] CHAN: Loading UTXO database from '/Users/dnldd/sh/node/data/simnet/utxodb'
2021-07-18 10:59:33.398 [INF] CHAN: UTXO database loaded
2021-07-18 10:59:33.400 [INF] INDX: Transaction index is enabled
2021-07-18 10:59:33.417 [INF] INDX: Address index is enabled
2021-07-18 10:59:33.417 [INF] INDX: Exists address index is enabled
2021-07-18 10:59:33.539 [DBG] FEES: Loaded fee estimator database
2021-07-18 10:59:33.544 [INF] CHAN: Loading block index...
2021-07-18 10:59:33.546 [DBG] CHAN: Block index loaded in 1.993375ms
2021-07-18 10:59:33.547 [INF] STKE: Stake database version 1 loaded
2021-07-18 10:59:33.548 [INF] CHAN: UTXO cache initializing (max size: 150 MiB)...
2021-07-18 10:59:33.548 [INF] CHAN: UTXO cache initialization completed
2021-07-18 10:59:33.551 [DBG] INDX: Current internal block ID: 20
2021-07-18 10:59:33.553 [INF] INDX: Removed 5 orphaned blocks from address index (heights 16 to 20)
2021-07-18 10:59:33.553 [ERR] DCRD: Unable to start server: assertion failed: dbIndexDisconnectBlock must be called with the block at the current index tip (transaction index, tip 6b591c0d4dc5055ba0a6f86ebdf89a38d2309563d97dce07a4262119e4a0ea81, block 1132ac204aaedab4db09dc8f2475eea81a9006fcfbc156dacf5ca25017dbe2da)
2021-07-18 10:59:33.553 [INF] DCRD: Gracefully shutting down the UTXO database...
2021-07-18 10:59:33.553 [INF] DCRD: Gracefully shutting down the block database...
2021-07-18 10:59:33.592 [INF] DCRD: Shutdown complete
```

dnldd commented:
Can confirm this is an issue on master as well, the requirement of passing the the index tip when disconnecting blocks cannot be met if the index is optional and can be disabled.

rstaudt2 commented:
This is the panic. The issue is that the index cannot use the spend journal to roll back since it's been removed.  You are correct, I just tested on master, and the issue exists there as well:

```
panic: missing spend journal data for 00000000000000001bd0c98c0fad436bd2b4650434927b437b9b840cf465aef8 [recovered]
	panic: missing spend journal data for 00000000000000001bd0c98c0fad436bd2b4650434927b437b9b840cf465aef8

goroutine 1 [running]:
github.com/decred/dcrd/database/v2/ffldb.rollbackOnPanic(0xc00017c870)
	/dcrd/database/ffldb/db.go:1891 +0x77
panic(0x170fb80, 0xc01661b4e0)
	/usr/local/opt/go/libexec/src/runtime/panic.go:969 +0x1b9
github.com/decred/dcrd/blockchain/v4.panicf(0x182bf40, 0x21, 0xc0001a2ab8, 0x1, 0x1)
	/dcrd/blockchain/chain.go:56 +0x134
github.com/decred/dcrd/blockchain/v4.dbFetchSpendJournalEntry(0x1a663c0, 0xc00017c870, 0xc01641da80, 0x1, 0x0, 0x100d80a, 0x174cae0, 0x17d2bc0, 0x1)
	/dcrd/blockchain/chainio.go:764 +0x745
github.com/decred/dcrd/blockchain/v4.(*chainQueryerAdapter).PrevScripts(0xc000b13630, 0x1a663c0, 0xc00017c870, 0xc01641da80, 0xc00020a380, 0x1, 0x0, 0x203000)
	/dcrd/blockchain/chain.go:2095 +0x89
github.com/decred/dcrd/blockchain/v4/indexers.(*Manager).Init.func2(0x1a663c0, 0xc00017c870, 0xc00017c870, 0x0)
	/dcrd/blockchain/indexers/manager.go:468 +0x37d
github.com/decred/dcrd/database/v2/ffldb.(*db).Update(0xc0001305c0, 0xc0004169c0, 0x0, 0x0)
	/dcrd/database/ffldb/db.go:1948 +0xa2
github.com/decred/dcrd/blockchain/v4/indexers.(*Manager).Init(0xc0004a6450, 0x1a5e3a0, 0xc000130300, 0x1a5fe60, 0xc000b13630, 0x0, 0x0)
	/dcrd/blockchain/indexers/manager.go:440 +0x402
github.com/decred/dcrd/blockchain/v4.New(0x1a5e3a0, 0xc000130300, 0xc0001a3560, 0x0, 0x42, 0xc00020a440)
	/dcrd/blockchain/chain.go:2293 +0x1024
main.newServer(0x1a5e3a0, 0xc000130300, 0xc00010f000, 0x1, 0x1, 0x1a617e0, 0xc0001305c0, 0x1a617e0, 0xc000130100, 0xc0000d4b00, ...)
	/dcrd/server.go:3419 +0xa9d
main.dcrdMain(0x0, 0x0)
	/dcrd/dcrd.go:207 +0xb68
main.main()
	/dcrd/dcrd.go:264 +0xca
```

rstaudt2 commented on blockchain/chain.go:
```diff
@@ -2099,6 +2102,31 @@ func (q *chainQueryerAdapter) PrevScripts(dbTx database.Tx, block *dcrutil.Block
+// provided block hash.
+//
+// This function is safe for concurrent access and is part of the
+// spendPruner.SpendPurger interface.
```
```suggestion
// spendpruner.SpendPurger interface.
```

rstaudt2 commented on blockchain/internal/spendpruner/error.go:
```diff
@@ -0,0 +1,54 @@
+	// ErrNeedSpendData indicates an error asserting a spend data dependency.
+	ErrNeedSpendData = ErrorKind("ErrNeedSpendData")
+
+	// ErrNeedSpendData indicates an error updating consumer spend dependencies.
```
```suggestion
	// ErrUpdateConsumerDeps indicates an error updating consumer spend dependencies.
```

rstaudt2 commented on blockchain/internal/spendpruner/pruner_test.go:
```diff
@@ -0,0 +1,346 @@
+	}
+
+	// Ensure removing a non-existent dependency does nothing.
+	pruner.RemoveSpendConsumerDependency(hashA, "consumer_x")
```
Should check for an error where `RemoveSpendConsumerDependency` is called (here and below) since it was updated to return an error.

rstaudt2 reviewed (commented):
I just had a few comments based on the updates, but otherwise, this looks good to me.  The panic should not be an issue for this PR since it exists on master and is addressed in #2219.

rstaudt2 reviewed (approved):
Looks good to me with the updates.  Nice work!

davecgh reviewed (changes requested):
Overall, this looks pretty close once the final things are addressed.  The biggest concern is the issue with the core logic being racy.  I have some more reviewing and testing to do, but I wanted to push this now to get discussion started on the race issue.

rstaudt2 commented on blockchain/chain.go:
```diff
@@ -703,6 +708,9 @@ func (b *BlockChain) connectBlock(node *blockNode, block, parent *dcrutil.Block,
 	b.stateLock.Unlock()
 
+	// Notify the spend pruner of the connected block.
+	go b.spendPruner.NotifyConnectedBlock(block.Hash())
```
Agreed with @davecgh's assessment here that this could result in a spend journal entry being removed for a block on the main chain, which would subsequently result in a [panic](https://github.com/decred/dcrd/blob/598bf66f24a671f90c0c1b3e190103159e2cf4ab/blockchain/chainio.go#L764) if that block needs to be disconnected.

davecgh replied:
To follow up, I was able to cause a failure randomly by modifying the code to intentionally add delays in just the right way so both events happen around the same time and just keep doing it until it eventually (because channel selection is random) hits the issue I described.  I suspect it could be reproduced a little more easily by using a separate mutex and synchronizing things in the "wrong" order such that it's constantly contending on the two channels, but I didn't spend any more time since I was able to reproduce and prove it is an issue.

I believe a way to fix things properly is to make two changes:

1) Combine `pruneCh` and `connectCh` into a single channel with a discriminant/type to specify which action is to be taken in order to ensure all modifications are properly ordered
2) Move the actual removal logic to a method on `blockchain` (that is provided to the pruner as a callback) such that said method takes the `processLock` which ensures things can't change out from under it.  Then, inside that function, only remove the spend entry when it is **not** part of the main chain and ignore it otherwise.

The combination of those two things will ensure a consistent state despite the indexing and pruning being done independently and asynchronously because former will ensure the pruner code itself will remove entries for the cases it knows about and the latter will do the same for the cases it does not.

dnldd replied:
Noted, thanks for the review.

rstaudt2 reviewed (changes requested):
Will re-review and re-test after the concurrency issue that @davecgh pointed out is addressed.  It would be useful to have an overview of the concurrency pattern being used with the updates as well to more thoroughly review from a design perspective (especially given that it is difficult to hit the scenarios that could cause issues in normal testing).

dnldd commented:
The latest commit adds these changes:

- This updates the spend pruner to use a single notification
channel to keep the order and processing of connectiong 
and disconnections deterministic.

- The process lock is now held when spend journal entries 
being removed by the pruner to ensure.

- The spend pruners lifecycle process uses an unbuffered 
channel to receive and process notifications. Notifications are
sent via goroutines to avoid blocking as a result.

- Comment related wrapping issues should be resolved.

davecgh commented on blockchain/internal/spendpruner/pruner.go:
```diff
@@ -0,0 +1,351 @@
+
+const (
+	// SPBlockConnected indicates a new block was connected to the chain.
+	SPBlockConnected spendPrunerEventType = iota
```
It's not a big deal since it's an internal package, but these should not be exported since they are only used internally.

dnldd replied:
noted, updating.

davecgh reviewed (changes requested):
I'll need to do some testing, but the updates look correct.

rstaudt2 reviewed (approved):
The updates look good to me.  Left some comments for non-consequential things inline.  I also re-tested this, including forcing `RemoveSpendEntry` to be called for a block that is part of the main chain to ensure that it does not remove the spend journal as intended.

dnldd commented on blockchain/internal/spendpruner/pruner.go:
```diff
@@ -10,46 +10,62 @@ import (
+type SpendJournalNotification struct {
+	BlockHash *chainhash.Hash
+	Event     spendPrunerEventType
+	Done      chan bool
```
good catch, was only focused on the test case I was updating.  will updated the connected case.

davecgh commented:
Please squash the changes in the review commit into the appropriate commit and I'll get this merged.
